### DISCUSSION TITLE
iroh 2.0 - What about WebAssembly?

### DISCUSSION THREAD

**Discussion by tennox** [n0-computer/iroh#787] [Ideas]
Dear iroh team,

I like the endavour to try to get IPFS to "just work [well]", and I'd like to bring attention & curiosity to our use-case:
**In-browser applications**.

We've wrestled a bit with trying to get rust stuff to work both on normal system and in-browser, so I know it's likely not part of the "absolute smallest thing that fits the majority of use cases." - but I'd love to know if it's part of your roadmap and/or what's needed for it.

**Comment by wtf-jik** [comment]
+1, imo it feels like a misstep for a library to not target wasm going forward.

**Reply by tennox** [reply]
Agreed. But their goal is also rather... huge. So I can see how this is not top of the list :sweat: 
But at least with rust & quic it's not totally off the table (I guess)

**Comment by fusetim** [comment]
~~I feel like iroh is mostly dependent on rust-libp2p, and therefore, wasm-target would need to be supported there first (on roadmap but still unplanned).~~ Apparently they are not? Did they reimplement all the libp2p stack?
Nonetheless, it could be nice to have :)

**Reply by Winterhuman** [reply]
@fusetim They mention it in https://number-zero.notion.site/Why-is-Iroh-moving-away-from-libp2p-57838618157f43c280b1fec381546604, they're moving away from Libp2p entirely to simplify things (and for other reasons, the Notion page explains it best).

From what I can tell, Iroh is targetted for server and desktop use, without the need to deal with browser restrictions, so it's unlikely WASM support will come anytime soon.

**Reply by b5** [reply]
We're not re-implementing the libp2p stack, we're cutting out a bunch of abstractions focusing mainly on QUIC & UDP. This does mean we still need access to raw UDP sockets. As mentioned in another reply, I'd like to see us ship a WASM library for handling the bits that _don't_ need low-level network acccess.

**Comment by b5** [comment]
WebAssembly is not the highest thing on our list of priorities, but it's also not the lowest.

We think browsers are an increasingly hostile environment for doing p2p stuff, so we're focused on native mobile. Targeting WebAssembly along with all of our other platforms for the entire iroh codebase is a bridge too far to make something that works.
_however_, we want to keep the consistent rule that _content is always validated at the edge_, which means instead of an HTTP gateway we want to ship an API that a WASM executable can use to integrate with an API (ideally that API uses WebTransport). We'll likely to delegate to a more capable node when necessary for NAT traversal stuff in the short term, but still do at least last mile data transfer and merkle verification from the browser.

In the long term, I'd very much like to see us push to add WebAssembly as a first-class platform, but as @tennox has said, we have a lot of ground to cover first
